### ISSUE TITLE
Support gRPC

### ISSUE THREAD

**Issue by doctor-q** [elastic/apm-agent-java#256]
We use rpc frameworks in our microservices, like thrift, grpc, and dubbo(alibaba)

**Comment by felixbarny** [comment]
Please also participate in the survey if you haven't already: https://docs.google.com/forms/d/e/1FAIpQLScd0RYiwZGrEuxykYkv9z8Hl3exx_LKCtjsqEo1OWx8BkLrOQ/viewform?usp=sf_link

**Comment by mariano-dm** [comment]
Hello. Thanks @giladgal for pointing me here.

We're a heavy java/python shop. as we're moving from rest to grpc, apps have been disappearing from our monitoring.  

having grpc support for the most popular frameworks in java would be a win.

protobuf is getting popular also

**Comment by peterbecker** [comment]
Shouldn't this be more than one ticket? I'm starting to look at GRPC support, but the thrift and dubbo are not relevant to us.

Protobuf is a different thing: it's a serialization format, not a wire protocol. GRPC builds on top of protobuf (it's messages are defined in protobuf format). It's a bit like REST/HTTP to JSON (if JSON would be the only choice).

**Comment by mariano-dm** [comment]
Correct, I was thinking about a service just using 'protobuf-java' lib, or dropwizard-grpc.

**Comment by peterbecker** [comment]
We are just using https://www.xolstice.org/protobuf-maven-plugin/ and write Kotlin code on top. My plan was to look at identifying how to pick up the implementations of the generated server stubs of that.

**Comment by peterbecker** [comment]
I started working on GRPC support at https://github.com/peterbecker/apm-agent-java/tree/grpc-plugin -- right now it's barely a test app, but at least I got a working protobuf/grpc setup.

My plan is to:

1. write test app (done)
2. write tests for test app to prove it actually works
3. write integration tests with APM agent (akin to e.g. the application server ones), failing on some basic things not being captured
4. start writing the plugin to make tests green
5. repeat steps 3 and 4 with more complex scenarios

Is that reasonable? For obvious reasons this is not an easy code base to get into, so far I've only used Java agents. I've found https://www.elastic.co/blog/a-cookbook-for-contributing-a-plugin-to-the-elastic-apm-java-agent -- but that doesn't cover the testing side as far as I can tell.
